quantum/lamparray.c

```diff
@@ -0,0 +1,191 @@
+// Copyright 2023 QMK
+// SPDX-License-Identifier: GPL-2.0-or-later
+#include "lamparray.h"
+#include "keycodes.h"
+#include "keymap_introspection.h"
+#include "action_layer.h"
+
+// Defaults are generated from info.json layout content
+#ifndef LAMPARRAY_WIDTH
+#    define LAMPARRAY_WIDTH ESTIMATED_KEYBOARD_WIDTH
+#endif
+#ifndef LAMPARRAY_HEIGHT
+#    define LAMPARRAY_HEIGHT ESTIMATED_KEYBOARD_HEIGHT
+#endif
+#ifndef LAMPARRAY_DEPTH
+#    define LAMPARRAY_DEPTH 30000
+#endif
+#ifndef LAMPARRAY_KIND
+#    define LAMPARRAY_KIND LAMPARRAY_KIND_KEYBOARD
+#endif
+
+// TODO: Implement backing API
+void lamparray_backing_enable(bool enable);
+void lamparray_backing_set_item(uint16_t index, lamp_state_t color);
+void lamparray_backing_update_finished(void);
+
+/**
+ * \brief Query a HID usage for a given location
+ *
+ * This can be requested while the user is changing layers. This is mitigated somewhat by assuming the default layer changes less frequently.
+ * This is currently accepted as a limitation as there is no method to invalidate the hosts view of the data.
+ */
+static inline uint16_t binding_at_keymap_location(uint8_t row, uint8_t col) {
+    uint16_t keycode = keycode_at_keymap_location(get_highest_layer(default_layer_state), row, col);
+#if LAMPARRAY_KIND == LAMPARRAY_KIND_KEYBOARD
+    // Basic QMK keycodes currently map directly to Keyboard UsagePage so safe to return without added indirection
+    // Mousekeys are ignored due to values overlap Keyboard UsagePage
+    if (IS_BASIC_KEYCODE(keycode) || IS_MODIFIER_KEYCODE(keycode)) {
+        return keycode;
+    }
+#elif LAMPARRAY_KIND == LAMPARRAY_KIND_MOUSE
+    // Usages from the Button UsagePage (0x09) in the range of Button1 (0x01) to Button5 (0x05) inclusive
+    if ((code) >= KC_MS_BTN1 && (code) <= KC_MS_BTN5) {
+        return keycode - KC_MS_BTN1;
+    }
+#endif
+    return 0;
+}
+
+#ifdef RGB_MATRIX_ENABLE
+#    include "rgb_matrix.h"
+
+#    define LAMPARRAY_RED_LEVELS 255
+#    define LAMPARRAY_GREEN_LEVELS 255
+#    define LAMPARRAY_BLUE_LEVELS 255
+#    define LAMPARRAY_INTENSITY_LEVELS 1
+#    define LAMPARRAY_LAMP_COUNT RGB_MATRIX_LED_COUNT
+#    define LAMPARRAY_UPDATE_INTERVAL (RGB_MATRIX_LED_FLUSH_LIMIT * 1000)
+
+/**
+ * \brief Get feature specific lamp info.
+ *
+ * Scales the LED config with the assumed RGB Matrix dimensions (224x64), for simplicity, as a completely flat device.
+ * Assumes all keys are either on the top or bottom of the resulting rectangle.
+ */
+__attribute__((weak)) void lamparray_get_lamp_info_data(uint16_t lamp_id, lamparray_attributes_response_t* data) {
+    data->position.x = (LAMPARRAY_WIDTH / 224) * g_led_config.point[lamp_id].x;
+    data->position.y = (LAMPARRAY_HEIGHT / 64) * (64 - g_led_config.point[lamp_id].y);
```

Comment:
The Y axis direction is actually the same in RGB Matrix and HID:
```suggestion
    data->position.y = (LAMPARRAY_HEIGHT / 64) * g_led_config.point[lamp_id].y;
```

Reply:
Does not match what i am seeing on multiple other devices.